src/controller.coffee

```diff
@@ -817,26 +822,57 @@ Editor::redo = ->
   currentValue = @getSerializedEditorState()
 
   until @redoStack.length is 0 or
-      (@redoStack[@redoStack.length - 1] is 'CAPTURE' and
+      (@redoStack[@redoStack.length - 1] instanceof CapturePoint and
       not @getSerializedEditorState().equals(currentValue))
     operation = @popRedo()
     if operation instanceof FloatingOperation
       @performFloatingOperation(operation, 'forward')
     else
       @getDocument(operation.document).perform(
-        operation.operation, 'forward', (if operation.document is @cursor.document then [@cursor.location] else [])
-      ) unless operation is 'CAPTURE'
+        operation.operation, 'forward', @getPreserves(operation.document)
+      ) unless operation instanceof CapturePoint
+
+  # Set the the remembered socket contents to the state it was in
+  # at this point in the undo stack.
+  if @undoStack[@undoStack.length - 1] instanceof CapturePoint
+    @rememberedSockets = @undoStack[@undoStack.length - 1].rememberedSockets.map (x) -> x.clone()
 
   @popRedo()
   @redrawMain()
   return
 
+# ## undoCapture and CapturePoint ##
+# A CapturePoint is a flag indicating that the undo stack
```

Comment:
... "is a sentiinal in the undo stock indicating the boundary of a user operation" ...